ledger/ledger_test.go

```diff
@@ -1943,6 +1944,198 @@ func TestLedgerReloadShrinkDeltas(t *testing.T) {
 	}
 }
 
+// TestLedgerMigrateV6ShrinkDeltas checks the ledger has correct account state
+// after performing a db migration from V6
+func TestLedgerMigrateV6ShrinkDeltas(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	accountDBVersion = 6
+	dbName := fmt.Sprintf("%s.%d", t.Name(), crypto.RandUint64())
+	testProtocolVersion := protocol.ConsensusVersion("test-protocol-migrate-shrink-deltas")
+	proto := config.Consensus[protocol.ConsensusV31]
+	proto.RewardsRateRefreshInterval = 500
+	config.Consensus[testProtocolVersion] = proto
+	defer func() {
+		delete(config.Consensus, testProtocolVersion)
+	}()
+	genesisInitState, initKeys := ledgertesting.GenerateInitState(t, testProtocolVersion, 10_000_000_000)
+	const inMem = false
+	cfg := config.GetDefaultLocal()
+	cfg.MaxAcctLookback = proto.MaxBalLookback
+	log := logging.TestingLog(t)
+	log.SetLevel(logging.Info) // prevent spamming with ledger.AddValidatedBlock debug message
+	dbs, _, err := openLedgerDB(dbName, inMem)
+	require.NoError(t, err)
+	err = dbs.Wdb.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		accountsCreateOnlineAccountsTable(ctx, tx)
+		accountsCreateTxTailTable(ctx, tx)
+		accountsCreateOnlineRoundParamsTable(ctx, tx)
+		return nil
+	})
+	l, err := OpenLedger(log, dbName, inMem, genesisInitState, cfg)
+	require.NoError(t, err)
+	defer func() {
+		l.Close()
+		os.Remove(dbName + ".block.sqlite")
+		os.Remove(dbName + ".tracker.sqlite")
+	}()
+
+	maxBlocks := 2000
+	accounts := make(map[basics.Address]basics.AccountData, len(genesisInitState.Accounts))
+	keys := make(map[basics.Address]*crypto.SignatureSecrets, len(initKeys))
+	// regular addresses: all init accounts minus pools
+	addresses := make([]basics.Address, len(genesisInitState.Accounts)-2, len(genesisInitState.Accounts))
+	i := 0
+	for addr := range genesisInitState.Accounts {
+		if addr != testPoolAddr && addr != testSinkAddr {
+			addresses[i] = addr
+			i++
+		}
+		accounts[addr] = genesisInitState.Accounts[addr]
+		keys[addr] = initKeys[addr]
+	}
+	sort.SliceStable(addresses, func(i, j int) bool { return bytes.Compare(addresses[i][:], addresses[j][:]) == -1 })
+
+	onlineTotals := make([]basics.MicroAlgos, maxBlocks+1)
+	curAddressIdx := 0
+	maxValidity := basics.Round(20) // some number different from number of txns in blocks
+	txnIDs := make(map[basics.Round]map[transactions.Txid]struct{})
+	// run for maxBlocks rounds with random payment transactions
+	// generate 1000 txn per block
+	for i := 0; i < maxBlocks; i++ {
+		stxns := make([]transactions.SignedTxn, 10)
```

Comment:
add a random number of transactions say from 7 to 15

Reply:
so random # btwen 7-15 instead of 10 each time?

Reply:
yes, please